File: java/77.组合.java
/*
 * @lc app=leetcode.cn id=77 lang=java
 *
 * [77] 组合
 */

// @lc code=start
class Solution {
    List<List<Integer>> res = new LinkedList<>();
    /**
     * 递归算法
     * 递归里面有个循环体，用来选取第一个数字
     * 23ms   45%   11%   41.3MB
     */
    public List<List<Integer>> combine(int n, int k) {
        if(k<=0 || n<k) return res;
        Deque<Integer> path = new LinkedList<>();
        dfs(n,k,1,path);

        return res;
    }
    private void dfs(int n, int k,int start, Deque<Integer> path){
        if(path.size() == k){
            res.add(new LinkedList<>(path));
            return;
        }
        for(int i=start;i<=n;i++){
        path.push(i);
        dfs(n,k,i+1,path);
        path.pop();
    }
    }
/**
 * 剪枝
 * 3ms   74%    7.9%   41.6%
 */
    private void dfs1(int n, int k,int start, Deque<Integer> path){
        if(path.size() == k){
            res.add(new LinkedList<>(path));
            return;
        }
        for(int i=start;i<=n-(k-path.size())+1;i++){
        path.push(i);
        dfs(n,k,i+1,path);
        path.pop();
    }
    }
}
// @lc code=end

File: py/77.组合.py
#
# @lc app=leetcode.cn id=77 lang=python3
#
# [77] 组合
#

# @lc code=start
class Solution:
    # 递归算法
    # 用循环选择第一个元素，然后用递归选择第二个元素
    # 404   61%   11.6%    14.9MB
    def combine1(self, n: int, k: int) -> List[List[int]]:
        res = []
        inner_list = []
        def dfs(start,count):
            if count == k:
                res.append(inner_list.copy())
                return
            for i in range(start+1,n+1):
                inner_list.append(i)
                dfs(i,count+1)
                inner_list.pop()
        dfs(0,0)
        return res
        # 在算法一的基础上优化了变量，增加了减枝
        # 64ms    73%     10.9%   14.9MB
    def combine(self, n: int, k: int) -> List[List[int]]:
        res = []
        inner_list = []
        def dfs(start):
            if len(inner_list) == k:
                res.append(inner_list.copy())
                return
            for i in range(start+1,n+len(inner_list)-k+2):
                inner_list.append(i)
                dfs(i)
                inner_list.pop()
        dfs(0)
        return res
# @lc code=end

